src/neo/SmartContract/Native/NameService.cs

```diff
@@ -0,0 +1,265 @@
+            if (engine.Snapshot.Storages.TryGet(CreateStorageKey(Prefix_Token).Add(hash)) is not null) return false;
+            StringList roots = engine.Snapshot.Storages.TryGet(CreateStorageKey(Prefix_Roots))?.GetInteroperable<StringList>();
+            if (roots is null || roots.BinarySearch(names[1]) < 0) throw new InvalidOperationException();
+            engine.AddGas(GetPrice(engine.Snapshot));
+            NameState state = new NameState
+            {
+                Owner = owner,
+                Name = name,
+                Description = "",
+                Expiration = (uint)(engine.Snapshot.PersistingBlock.Timestamp / 1000) + OneYear
+            };
+            Mint(engine, state);
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_Expiration).AddBigEndian(state.Expiration).Add(hash), new StorageItem(new byte[] { 0 }));
+            return true;
+        }
+
+        [ContractMethod(0, CallFlags.WriteStates)]
+        private uint Renew(ApplicationEngine engine, string name)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            string[] names = name.Split('.');
+            if (names.Length != 2) throw new ArgumentException(null, nameof(name));
+            engine.AddGas(GetPrice(engine.Snapshot));
+            byte[] hash = GetKey(Utility.StrictUTF8.GetBytes(name));
+            NameState state = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Token).Add(hash)).GetInteroperable<NameState>();
+            engine.Snapshot.Storages.Delete(CreateStorageKey(Prefix_Expiration).AddBigEndian(state.Expiration).Add(hash));
+            state.Expiration += OneYear;
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_Expiration).AddBigEndian(state.Expiration).Add(hash), new StorageItem(new byte[] { 0 }));
+            return state.Expiration;
+        }
+
+        [ContractMethod(0_03000000, CallFlags.WriteStates)]
+        private void SetAdmin(ApplicationEngine engine, string name, UInt160 admin)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            string[] names = name.Split('.');
+            if (names.Length != 2) throw new ArgumentException(null, nameof(name));
+            if (!engine.CheckWitnessInternal(admin)) throw new InvalidOperationException();
+            NameState state = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_Token).Add(GetKey(Utility.StrictUTF8.GetBytes(name)))).GetInteroperable<NameState>();
+            if (!engine.CheckWitnessInternal(state.Owner)) throw new InvalidOperationException();
+            state.Admin = admin;
+        }
+
+        [ContractMethod(0_30000000, CallFlags.WriteStates)]
+        private void SetRecord(ApplicationEngine engine, string name, RecordType type, string data)
+        {
+            if (!nameRegex.IsMatch(name)) throw new ArgumentException(null, nameof(name));
+            switch (type)
```

Comment:
Allow more or less records according to the payment?

Reply:
What do you mean?

Reply:
If the user pay X he only can store Y records. If he want more, he should pay more, it's just an idea.

Reply:
But he need to pay for `SetRecord` too.